runtime/src/iree/hal/drivers/cuda/tracing.c

```diff
@@ -169,98 +195,188 @@ void iree_hal_cuda_tracing_context_free(
+    while (event) {
+      uint32_t query_id = (uint32_t)(event - &context->event_pool[0]);
+
+      CUresult result = context->symbols->cuEventQuery(event->event);
+      while (result == CUDA_ERROR_NOT_READY) {
+        result = context->symbols->cuEventQuery(event->event);
+      }
       if (result != CUDA_SUCCESS) break;
 
       // Calculate context-relative time and notify tracy.
       float relative_millis = 0.0f;
       IREE_CUDA_IGNORE_ERROR(
           context->symbols,
           cuEventElapsedTime(&relative_millis, context->base_event,
-                             query_event));
+                             event->event));
       int64_t gpu_timestamp = (int64_t)((double)relative_millis * 1000000.0);
       iree_tracing_gpu_zone_notify(context->id, query_id, gpu_timestamp);
 
-      read_query_count = i + 1;
+      read_query_count += 1;
+      event = event->next_in_cb;
     }
-    IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, (int64_t)read_query_count);
+    iree_hal_cuda_tracing_context_event_t* next = events->next_submission;
+    events->next_submission = events;
+    events = next;
+    context->event_submitted_list_head = events;
+  }
+  IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, (int64_t)read_query_count);
+  IREE_TRACE_ZONE_END(z0);
+  iree_slim_mutex_unlock(&context->event_mutex);
+}
+
+void iree_hal_cuda_tracing_notify_submitted(
+    iree_hal_cuda_tracing_context_t* context,
+    iree_hal_cuda_tracing_context_event_t** event_list_begin,
+    iree_hal_cuda_tracing_context_event_t** event_list_end) {
+  if (!context) return;
+  iree_slim_mutex_lock(&context->event_mutex);
+
+  IREE_ASSERT_ARGUMENT(event_list_begin);
+  IREE_ASSERT_ARGUMENT(event_list_end);
 
-    context->query_tail += read_query_count;
-    if (context->query_tail >= context->query_capacity) {
-      context->query_tail = 0;
+  if (!*event_list_begin) {
+    iree_slim_mutex_unlock(&context->event_mutex);
+    return;
+  }
+
+  iree_hal_cuda_tracing_context_event_t* evt = *event_list_begin;
+  while (evt) {
+    evt = evt->next_in_cb;
+  }
+
+  if (!context->event_submitted_list_head) {
+    context->event_submitted_list_head = *event_list_begin;
+    context->event_submitted_list_tail = *event_list_begin;
+    iree_slim_mutex_unlock(&context->event_mutex);
+    return;
+  }
+
+  context->event_submitted_list_tail->next_submission = *event_list_begin;
+  context->event_submitted_list_tail = *event_list_begin;
+  iree_slim_mutex_unlock(&context->event_mutex);
+}
+
+void iree_hal_cuda_tracing_free(
+    iree_hal_cuda_tracing_context_t* context,
+    iree_hal_cuda_tracing_context_event_t** event_list_begin,
+    iree_hal_cuda_tracing_context_event_t** event_list_end) {
+  if (!context) return;
+  iree_slim_mutex_lock(&context->event_mutex);
+
+  IREE_ASSERT_ARGUMENT(event_list_begin);
+  IREE_ASSERT_ARGUMENT(event_list_end);
+
+  if (!*event_list_begin) {
+    iree_slim_mutex_unlock(&context->event_mutex);
+    return;
+  }
+
```

Comment:
Maybe putting some comment here explain the following overall? Something like "We are done with this event list. Perform final work with Tracy and return to the context's freelist".